src/Google.Bigquery.V2/SchemaBuilder.cs

```diff
@@ -0,0 +1,170 @@
+        /// <param name="mode">The mode of the field. Must be a defined member within <see cref="SchemaFieldMode"/>.</param>
+        /// <param name="description">The description of the field. May be null.</param>
+        public void Add(string name, TableSchema nestedSchema, SchemaFieldMode mode = SchemaFieldMode.Nullable, string description = null)
+        {
+            Preconditions.CheckNotNull(name, nameof(name));
+            Preconditions.CheckNotNull(nestedSchema, nameof(nestedSchema));
+            Preconditions.CheckArgument(Enum.IsDefined(typeof(SchemaFieldMode), mode), nameof(mode), "Specified value is not a member of SchemaFieldMode");
+
+            Add(new TableFieldSchema
+            {
+                Name = name,
+                Type = "RECORD",
+                Fields = nestedSchema.Fields,
+                Mode = mode.ToString().ToUpperInvariant(),
+                Description = description,
+            });
+        }
+
+        /// <summary>
+        /// Creates a nested field with the specified schema, and adds it to the schema being built.
+        /// This method simply delegates to <see cref="Add(string, TableSchema, SchemaFieldMode, string)"/> after calling <see cref="Build"/>
+        /// on <paramref name="nestedSchema"/>.
+        /// </summary>
+        /// <param name="name">The name of the field. Must not be null.</param>
+        /// <param name="nestedSchema">The schema for the nested field, in the form of a <see cref="SchemaBuilder"/>. Must not be null.</param>
+        /// <param name="mode">The mode of the field. Must be a defined member within <see cref="SchemaFieldMode"/>.</param>
+        /// <param name="description">The description of the field. May be null.</param>
+        public void Add(string name, SchemaBuilder nestedSchema, SchemaFieldMode mode = SchemaFieldMode.Nullable, string description = null) =>
+            Add(Preconditions.CheckNotNull(name, nameof(name)),
+                Preconditions.CheckNotNull(nestedSchema, nameof(nestedSchema)).Build(),
+                mode,
+                description);
+
+        /// <summary>
+        /// Constructs a <see cref="TableSchema"/> from this builder.
+        /// </summary>
+        /// <returns>A <see cref="TableSchema"/> from the fields added to this builder.</returns>
+        public TableSchema Build() => new TableSchema { Fields = fields };
+
+
+        /// <summary>
+        /// Returns an iterator over the fields in this builder.
+        /// </summary>
+        /// <remarks>This method only exists to satisfy the C# language rules on collection initializers.</remarks>
+        /// <returns>An iterator over the fields in this builder.</returns>
+        IEnumerator IEnumerable.GetEnumerator() => fields.GetEnumerator();
+    }
+
+    /// <summary>
+    /// The type of a field within a schema.
+    /// </summary>
+    public enum SchemaFieldType
```

Comment:
Leaving deck-chairs for the moment :)
I think we're likely to want to shift schema things around a bit in general.